apps/passport-client/src/dispatch.ts

```diff
@@ -758,18 +776,24 @@ async function doSync(
   );
   if (state.serverStorageHash !== appStorage.storageHash) {
     console.log("[SYNC] sync action: upload");
-    // TODO(artwyman): Add serverStorageRevision input as knownRevision here,
-    // but only after we're able to respond to a conflict by downloading.
     const upRes = await uploadSerializedStorage(
       appStorage.serializedStorage,
-      appStorage.storageHash
+      appStorage.storageHash,
+      state.serverStorageRevision
     );
     if (upRes.success) {
       return {
         serverStorageRevision: upRes.value.revision,
         serverStorageHash: upRes.value.storageHash
       };
+    } else if (upRes.error.name === "Conflict") {
+      // Conflicts are resolved at download time, so ensure another download.
+      return {
+        completedFirstSync: true,
```

Comment:
another thing I've forgotten and would probably have to re-read the code to understand perfectly: under what conditions is `completedFirstSync` `true`? it could be a number of situations. worth commenting at the declaration in `state.ts` imo